integrations/anthropic/src/actions/generate-content.ts

```diff
@@ -0,0 +1,261 @@
+import Anthropic from '@anthropic-ai/sdk'
+import { MessageCreateParams } from '@anthropic-ai/sdk/resources/messages'
+import { llm } from '@botpress/common'
+import { InvalidPayloadError } from '@botpress/client'
+import { IntegrationLogger } from '@botpress/sdk/dist/integration/logger'
+
+type ModelSpecs = llm.types.ModelCost & {
+  /**
+   * Maximum number of output tokens supported by the model.
+   */
+  outputTokensLimit: number
+}
+
+export async function generateContent<M extends string>(
+  input: llm.schemas.GenerateContentInput,
+  anthropic: Anthropic,
+  logger: IntegrationLogger,
+  params: {
+    models: {
+      [key in M]: ModelSpecs
+    }
+  }
+): Promise<llm.schemas.GenerateContentOutput> {
+  const modelSpecs = params.models[input.model as M]
+  if (!modelSpecs) {
+    throw new InvalidPayloadError(
+      `Model name "${input.model}" is not allowed, supported model names are: ${Object.keys(params.models).join(', ')}`
+    )
+  }
+
+  if (input.responseFormat === 'json_object') {
+    input.systemPrompt =
+      (input.systemPrompt || '') +
+      '\n\nYour response must always be in valid JSON format and expressed as a JSON object.'
+  }
+
+  const messages: Anthropic.MessageParam[] = []
+
+  for (const message of input.messages) {
+    messages.push(await mapToAnthropicMessage(message))
+  }
+
+  let response: Anthropic.Messages.Message
+
+  try {
+    response = await anthropic.messages.create({
+      model: input.model,
+      max_tokens: input.maxTokens || modelSpecs.outputTokensLimit,
+      temperature: input.temperature,
+      top_p: input.topP,
+      system: input.systemPrompt,
+      stop_sequences: input.stopSequences,
+      metadata: {
+        user_id: input.userId,
+      },
+      tools: mapToAnthropicTools(input),
+      tool_choice: mapToAnthropicToolChoice(input.toolChoice),
+      messages,
+    })
+  } catch (err) {
+    logger.forBot().error(`generateContent action failed because Anthropic returned an error: ${JSON.stringify(err)}`)
```

Comment:
:thought_balloon: **thought:** just wondering if we should try to scrub the api key from the error message before sending it to the user as-is, pretty sure it wouldn't be outputted but we never know

Reply:
It would be rare that this happened but you make a good point, I'm realizing this is serializing an arbitrary object here and we cannot really trust what this will log. I'll refactor this to instead extract the most relevant attributes of the error that can be exposed in the logs.

Reply:
Addressed in https://github.com/botpress/botpress/pull/13159/commits/cffe61aa70cae276cecb09d8a3d7dfb8a4c90499